src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs

```diff
@@ -149,5 +209,210 @@ internal static class SemanticClassificationUtilities
                 throw ExceptionUtilities.Unreachable;
             }
         }
+
+        private async Task AddSemanticClassificationsAsync(
+            IClassificationService classificationService,
+            Document document,
+            TextSpan textSpan,
+            List<ClassifiedSpan> classifiedSpans,
+            CancellationToken cancellationToken)
+        {
+            var workspace = document.Project.Solution.Workspace;
+            var fullyLoadedStateTask = s_workspaceToFullyLoadedStateTask.GetValue(
+                workspace, w =>
+                {
+                    var workspaceLoadedService = workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+                    return workspaceLoadedService.WaitUntilFullyLoadedAsync(CancellationToken.None);
+                });
+
+            // If we're not fully loaded try to read from the cache instead so that classifications appear up to date.
+            // New code will not be semantically classified, but will eventually when the project fully loads.
+            var isFullyLoaded = fullyLoadedStateTask.IsCompleted;
+            if (await TryAddSemanticClassificationsFromCacheAsync(document, textSpan, classifiedSpans, isFullyLoaded, cancellationToken).ConfigureAwait(false))
+                return;
+
+            await classificationService.AddSemanticClassificationsAsync(
+                document, textSpan, classifiedSpans, cancellationToken).ConfigureAwait(false);
+
+            // Once fully loaded, add this document to the work queue to persist its classifications to storage in the
+            // future.  This way we can store the full accurate set of classifications for it to help speed things up in
+            // VS on next launch.
+            //
+            // Only bother to do this for real documents that are being classified, not documents for things like
+            // ephemeral preview workspaces.
+            if (isFullyLoaded && document.Project.Solution.Workspace.Kind == WorkspaceKind.Host)
```

Comment:
The check for WorkspaceKind.Host here might backfire; we may have to tweak it but don't worry about fixing this until the PR leaves draft.